### PR TITLE
[Workflow] Add purge/pause/resume/raise event to docs

### PR DESCRIPTION
**Pull request by hhunter-ms**
## Description

Add purge/pause/resume API methods to documentation:
- Workflow API doc
- How to manage workflows doc

Need:
Response body examples for Workflow API doc

## Issue reference

PR will close: #3316 #3350 #3349

### CODE REVIEW

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 1]
```diff
@@ -56,6 +56,29 @@ To terminate your workflow with an ID `12345678`, run:
 POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/terminate
 ```
 
+### Pause or resume a workflow
```
Comment (RyanLettieri) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md:82]: @cgillum Correct me if I'm wrong, but for pause and resume we no longer need workflow names, right? So the pause and resume calls below can omit the "OrderProcessingWorkflow" and simply use the workflow component and instance ID. In fact, it looks like you already have that in the document below.

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 2]
```diff
@@ -90,6 +90,114 @@ Traceparent: 00-e3dedffedbeb9efbde9fbed3f8e2d8-5f38960d43d24e98-01
...
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceId>/pause
+```
+
+### URL parameters
+
+Parameter | Description
+--------- | -----------
+`workflowComponentName` | Current default is `dapr` for Dapr Workflows
+`workflowName` | Identify the workflow type
```
Comment (RyanLettieri) [inline daprdocs/content/en/reference/api/workflow_api.md]: As per the comment above, I don't believe workflowName is necessary here nor the url parameters tables below. This also applies to Purge

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 3]
```diff
@@ -90,6 +90,112 @@ Traceparent: 00-e3dedffedbeb9efbde9fbed3f8e2d8-5f38960d43d24e98-01
...
+---- | -----------
+`202`  | Accepted
+`400`  | Request was malformed
+`500`  | Request formatted correctly, error in dapr code or underlying component
+
+### Response content
+
+The API call will provide a response similar to this:
+
+```bash
```
Comment (hhunter-ms) [inline daprdocs/content/en/reference/api/workflow_api.md]: @RyanLettieri @cgillum I used the responses for "terminate workflow" as placeholders for pause/resume/purge - can you review and let me know what needs to be changed?

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 4]
```diff
@@ -30,7 +30,19 @@ WorkflowReference startResponse = await daprClient.StartWorkflowAsync(orderId, w
 GetWorkflowResponse getResponse = await daprClient.GetWorkflowAsync(orderId, workflowComponent, workflowName);
 
 // Terminate the workflow
-await daprClient.TerminateWorkflowAsync(instanceId, workflowComponent);
+GetWorkflowResponse getResponse = await daprClient.TerminateWorkflowAsync(orderId, workflowComponent, cts);
```
Comment (RyanLettieri) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: There is no response object returned from the terminate call. It should look like this:
await daprClient.TerminateWorkflowAsync(orderId, workflowComponent, cts);

This applies to the RaiseEvent, Pause, Resume, and Purge calls below as well

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 5]
```diff
@@ -58,6 +70,41 @@ To terminate your workflow with an ID `12345678`, run:
 POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/terminate
 ```
 
+### Raise an event
+
+For your workflow to create "wait for external event" tasks that:
+
+1. Subscribe to external events.
+1. Await those tasks to block execution until the event is received.
```
Comment (RyanLettieri) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Nit: Should be a 2 instead of a 1

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 6]
```diff
@@ -75,6 +75,114 @@ Code | Description
 
 This API does not return any content.
 
+## Raise Event workflow request
```
Comment (RyanLettieri) [inline daprdocs/content/en/reference/api/workflow_api.md]: Maybe we should reword this to say "Raise Event request for a workflow" or simply "Raise Event request"

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 7]
```diff
@@ -75,6 +75,114 @@ Code | Description
...
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
+```
+
+### URL parameters
+
+Parameter | Description
+--------- | -----------
+`workflowComponentName` | Current default is `dapr` for Dapr Workflows
+`instanceId` | Unique value created for each run of a specific workflow
+`eventName` | The name of any workflow function
```
Comment (RyanLettieri) [inline daprdocs/content/en/reference/api/workflow_api.md]: Reword " The name of any workflow function" to "The name of the event to raise"

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 5 above)
Comment (hhunter-ms) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: In markdown, using `1.` for all numbers in a numbered list translates to `1.`, `2.`, `3.`, etc. once published, without you having to remember to renumber the list if you add steps later.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 8]
```diff
@@ -58,6 +70,41 @@ To terminate your workflow with an ID `12345678`, run:
...
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/resume
+```
+
+### Purge a workflow 
+
+To remove actor state from your state store and avoid overloading your state stores, run:
+
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/dapr/OrderProcessingWorkflow/purge
```
Comment (RyanLettieri) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: "OrderProcessingWorkflow" should be "12345678" since we purge based on the instance ID, not the workflow name.

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 9]
```diff
@@ -75,6 +75,114 @@ Code | Description
...
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceId>/purge
+```
+
+### URL parameters
+
+Parameter | Description
+--------- | -----------
+`workflowComponentName` | Current default is `dapr` for Dapr Workflows
+`workflowName` | Identify the workflow type
```
Comment (RyanLettieri) [inline daprdocs/content/en/reference/api/workflow_api.md]: workflow name is not needed

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 10]
```diff
@@ -75,6 +75,114 @@ Code | Description
...
+`400`  | Request was malformed
+`500`  | Request formatted correctly, error in dapr code or underlying component
+
+### Response content
+
+None.
+
+## Pause workflow request
+
+Pause a running workflow instance with the given name and instance ID.
```
Comment (RyanLettieri) [inline daprdocs/content/en/reference/api/workflow_api.md]: Nit: Here and below you are saying "with the given name". Since these methods don't require a workflowname, what is this referring to?

Comment (hhunter-ms) [reply daprdocs/content/en/reference/api/workflow_api.md]: most likely it's just me copy/pasting and not double checking :) thanks for the catch!

**Comment by RyanLettieri** [comment]
LGTM

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 11]
```diff
@@ -23,14 +23,26 @@ OrderPayload input = new OrderPayload("Paperclips", 99.95);
 Dictionary<string, string> workflowOptions; // This is an optional parameter
 CancellationToken cts = CancellationToken.None;
 
-// Start the workflow. This returns back a "WorkflowReference" which contains the instanceID for the particular workflow instance.
-WorkflowReference startResponse = await daprClient.StartWorkflowAsync(orderId, workflowComponent, workflowName, input, workflowOptions, cts);
+// Start the workflow. This returns back a "StartWorkflowResponse" which contains the instanceID for the particular workflow instance.
+StartWorkflowResponse startResponse = await daprClient.StartWorkflowAsync(orderId, workflowComponent, workflowName, input, workflowOptions, cts);
 
 // Get information on the workflow. This response will contain information such as the status of the workflow, when it started, and more!
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]:
```suggestion
// Get information on the workflow. This response contains information such as the status of the workflow, when it started, and more!
```

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-author-workflow.md [hunk 12]
```diff
@@ -28,19 +28,86 @@ The Dapr sidecar doesn’t load any workflow definitions. Rather, the sidecar si
 
 ## Write the workflow activities
 
-Define the workflow activities you'd like your workflow to perform. Activities are a class definition and can take inputs and outputs. Activities also participate in dependency injection, like binding to a Dapr client.
+Define the workflow activities you'd like your workflow to perform. Activities are a class definition and can take inputs and outputs. Activities also participate in dependency injection, like binding to a Dapr client.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-author-workflow.md]: Does this paragraph apply to all SDKs? Go is not class orientated and will Go/JS/Python SDKs have the concept of dependency injection? Asking since this may be best moved into the .NET section and a more general statement added for what an activity is.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 13]
```diff
@@ -23,14 +23,26 @@ OrderPayload input = new OrderPayload("Paperclips", 99.95);
...
 
 // Terminate the workflow
-await daprClient.TerminateWorkflowAsync(instanceId, workflowComponent);
+await daprClient.TerminateWorkflowAsync(orderId, workflowComponent, cts);
+
+// Raise an event (an incoming purchase order) that your workflow will wait for. This returns the item waiting to be purchased.
+await daprClient.RaiseWorkflowEventAsync(orderId, workflowComponent, workflowName, input, cts);
+
+// Pause
+await daprClient.PauseWorkflowAsync(orderId, workflowComponent, cts);
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: I saw a discussion and code that said this was going to be called suspend, instead of pause. This code https://github.com/dapr/dapr/pull/6106/files. Is that correct?

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 14]
```diff
@@ -58,6 +70,41 @@ To terminate your workflow with an ID `12345678`, run:
...
+1. Subscribe to external events.
+1. Await those tasks to block execution until the event is received.
+
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
+```
+
+> An `eventName` can be any function. 
+
+### Pause or resume a workflow
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md:82]: Suspend instead of Pause? Which term will be used?

```suggestion
### Pause or resume a workflow
```

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 15]
```diff
@@ -58,6 +70,41 @@ To terminate your workflow with an ID `12345678`, run:
...
+
+To resume a workflow with an ID `12345678`, run:
+
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/resume
+```
+
+### Purge a workflow 
+
+To remove actor state from your state store and avoid overloading your state stores, run:
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Actors are an implementation detail and I suggest just saying remove the state. 
```suggestion
To remove all the workflow and activity state from your state store to reduce storage use, run:
```
Also when should you call purge? What is the recommendation, once a workflow has completed? SInce there is a danger here that if you call this whilst a workflow is running, won't this invalidate the workflow? What is the guidance here? Is there a "Danger: Be careful when calling purge API" warning?

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 16]
```diff
@@ -58,6 +70,41 @@ To terminate your workflow with an ID `12345678`, run:
 POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/terminate
 ```
 
+### Raise an event
+
+Run the following for your workflow to create "wait for external event" tasks that:
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: I still do not understand this. Is there an API to register for an event and an API to raise and event?
Can you provide a complete example of registering for an event X and then the workflow being triggered by event X happening? Reading this, it seems raise event is more like registered for an event. Is that correct, or am  I misunderstanding this?

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 17]
```diff
@@ -75,6 +75,113 @@ Code | Description
 
 This API does not return any content.
 
+## Raise Event request
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/workflow_api.md:78]: Raise and event or register for this event to be raised?

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 18]
```diff
@@ -75,6 +75,113 @@ Code | Description
...
+---- | -----------
+`202`  | Accepted
+`400`  | Request was malformed
+`500`  | Request formatted correctly, error in dapr code or underlying component
+
+### Response content
+
+None.
+
+## Pause workflow request
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/workflow_api.md:111]: Suspend or Pause?

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 19]
```diff
@@ -75,6 +75,113 @@ Code | Description
...
+
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceId>/pause
+```
+
+### URL parameters
+
+Parameter | Description
+--------- | -----------
+`workflowComponentName` | Current default is `dapr` for Dapr Workflows
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/workflow_api.md]:
```suggestion
`workflowComponentName` | Use `dapr` for Dapr Workflows
```

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 20]
```diff
@@ -75,6 +75,113 @@ Code | Description
...
+
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceId>/resume
+```
+
+### URL parameters
+
+Parameter | Description
+--------- | -----------
+`workflowComponentName` | Current default is `dapr` for Dapr Workflows
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/workflow_api.md]:
```suggestion
`workflowComponentName` | Use `dapr` for Dapr Workflows
```

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 21]
```diff
@@ -75,6 +75,113 @@ Code | Description
...
+
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceId>/purge
+```
+
+### URL parameters
+
+Parameter | Description
+--------- | -----------
+`workflowComponentName` | Current default is `dapr` for Dapr Workflows
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/workflow_api.md]:
```suggestion
`workflowComponentName` | Use `dapr` for Dapr Workflows
```

**Review by msfussell — COMMENTED** [review]
Here are my comments

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 13 above)
Comment (hhunter-ms) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: @RyanLettieri @cgillum can you confirm?

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 15 above)
Comment (hhunter-ms) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: @RyanLettieri could you elaborate the risks of calling purge?

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 16 above)
Comment (hhunter-ms) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: @RyanLettieri any suggestions for what is the best way to demonstrate the HTTP raise event call better?

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 13 above)
Comment (RyanLettieri) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: It was my understanding that the code was going to reflect Pause/Resume rather than Suspend/Resume. However, the value returned to the user for the status when they call GET on a workflow instance that has been paused is 'Suspended'. I think we should leave it as is for now and potentially revisit later on.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 15 above)
Comment (RyanLettieri) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: If the current runtime state is not completed, it is not able to be purged and will return an error stating "orchestration has not yet completed". Purge should be called on completed workflow instances to free up potentially non-needed items in the statestore.

File: daprdocs/content/en/reference/api/workflow_api.md (hunk 17 above)
Comment (RyanLettieri) [reply daprdocs/content/en/reference/api/workflow_api.md:78]: Raise an event

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 16 above)
Comment (RyanLettieri) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Hmm I don't think we can actually register an event through the HTTP api, only call an already registered event. Just to double check, @cgillum , is this correct?

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-author-workflow.md [hunk 22]
```diff
@@ -73,19 +142,21 @@ It also includes a `RunAsync` method that does the heavy lifting of the workflow
...
 
-Compose the workflow activities into a workflow. 
+Finally, compose the application using the workflow.
 
 {{< tabs ".NET" >}}
 
 {{% codetab %}}
 
-[In the following example](https://github.com/dapr/dotnet-sdk/blob/master/examples/Workflow/WorkflowConsoleApp/Program.cs), for a basic ASP.NET order processing application using the .NET SDK, your project code would include:
+[In the following `Project.cs` example](https://github.com/dapr/dotnet-sdk/blob/master/examples/Workflow/WorkflowConsoleApp/Program.cs), for a basic ASP.NET order processing application using the .NET SDK, your project code would include:
```
Comment (cgillum) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-author-workflow.md]:
```suggestion
[In the following `Program.cs` example](https://github.com/dapr/dotnet-sdk/blob/master/examples/Workflow/WorkflowConsoleApp/Program.cs), for a basic ASP.NET order processing application using the .NET SDK, your project code would include:
```

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 23]
```diff
@@ -23,14 +23,26 @@ OrderPayload input = new OrderPayload("Paperclips", 99.95);
 Dictionary<string, string> workflowOptions; // This is an optional parameter
 CancellationToken cts = CancellationToken.None;
 
-// Start the workflow. This returns back a "WorkflowReference" which contains the instanceID for the particular workflow instance.
-WorkflowReference startResponse = await daprClient.StartWorkflowAsync(orderId, workflowComponent, workflowName, input, workflowOptions, cts);
+// Start the workflow. This returns back a "StartWorkflowResponse" which contains the instanceID for the particular workflow instance.
```
Comment (cgillum) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]:
```suggestion
// Start the workflow. This returns back a "StartWorkflowResponse" which contains the instance ID for the particular workflow instance.
```

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 24]
```diff
@@ -23,14 +23,26 @@ OrderPayload input = new OrderPayload("Paperclips", 99.95);
 Dictionary<string, string> workflowOptions; // This is an optional parameter
 CancellationToken cts = CancellationToken.None;
```
Comment (cgillum) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Let's remove this `cts` variable and all references to it below. I've made a similar change already in [this PR](https://github.com/dapr/dotnet-sdk/pull/1087/files#diff-e3110f048c2796ea0bc661b963ead4b5f91d27ca1cffea9929ea614b8b8857cf).

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 13 above)
Comment (cgillum) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Unfortunately, there's an inconsistency which I think we missed in the implementation. The APIs are all called "Pause" but the workflow status is changed to "Suspended", as Ryan mentioned.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 16 above)
Comment (cgillum) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: This reminds me that we need to update the workflow patterns document to describe the human interaction pattern, which is where external events comes into play. Until then, the best docs we have right now are [here](https://docs.dapr.io/developing-applications/building-blocks/workflow/workflow-features-concepts/#external-events).

To answer the question, registering interest in an event is an engine-specific action. For example, in Dapr Workflow, your workflow can call a `WaitForExternalEvent` API to register interest in an event. Temporal has something similar. We don't have a common interface for registering interest in events - only a public API for raising the events.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 14 above)
Comment (cgillum) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md:82]: Pause is the term we're using.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 25]
```diff
@@ -58,6 +70,43 @@ To terminate your workflow with an ID `12345678`, run:
...
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
+```
+
+> An `eventName` can be any function. 
+
+### Pause or resume a workflow
+
+To plan for down-time, wait for inputs, and more, you can pause and then resume a workflow. To stop a workflow with an ID `12345678` until triggered to resume, run:
+
+```bash
```
Comment (cgillum) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Instead of `bash` I think it would be more appropriate to use `http` for all these.

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 26]
```diff
@@ -75,6 +75,113 @@ Code | Description
...
+`workflowComponentName` | Use `dapr` for Dapr Workflows
+`instanceId` | Unique value created for each run of a specific workflow
+
+### HTTP response codes
+
+Code | Description
+---- | -----------
+`202`  | Accepted
+`400`  | Request was malformed
+`500`  | Request formatted correctly, error in dapr code or underlying component
```
Comment (cgillum) [inline daprdocs/content/en/reference/api/workflow_api.md]: 500 doesn't necessarily mean the request was formatted correctly, so I'd take that part out.
```suggestion
`500`  | Error in Dapr code or underlying component
```

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 27]
```diff
@@ -75,6 +75,113 @@ Code | Description
...
+
+### Response content
+
+None.
+
+## Purge workflow request
+
+Purge the actor state from your state store with the workflow's instance ID.
```
Comment (cgillum) [inline daprdocs/content/en/reference/api/workflow_api.md]: Actors are an implementation detail that only applies to the Dapr Workflow engine. It's not applicable for other workflow engines. We should say we're purging the *workflow* state.
```suggestion
Purge the workflow state from your state store with the workflow's instance ID.
```

**Review by cgillum — COMMENTED** [review]
Added my review comments, and responded to a few comment threads.

**Comment by hhunter-ms** [comment]
@cgillum [for your comment re: Suspend vs Pause](https://github.com/dapr/docs/pull/3332#discussion_r1191482086)... should I update it to this then?

```
// Pause
await daprClient.SuspendOrchestration(orderId, workflowComponent);
```

**Comment by cgillum** [comment]
@hhunter-ms it should be `PauseWorkflowAsync`:
```csharp
// Pause
await daprClient.PauseWorkflowAsync(orderId, workflowComponent);
```

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 13 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Ok. So can we fix this issue in 1.11?

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 28]
```diff
@@ -58,6 +70,43 @@ To terminate your workflow with an ID `12345678`, run:
...
+```
+
+### Purge a workflow 
+
+To remove all the workflow and activity state from your state store to reduce storage use, run:
+
+```bash
+POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/purge
+```
+
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md:105]:
```suggestion
If the workflow state is not anything other than "Completed" calling `purge` returns an error stating "Workflow has not yet completed". Purge should only be called on completed workflow instances to free resources.
```

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 15 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: I added a comment to add this detail

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 28 above)
Comment (cgillum) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md:105]: I described this a bit more explicitly in the .NET SDK. How about this?

> The purge API can be used to permanently delete workflow metadata from the underlying state store,
> including any stored inputs, outputs, and workflow history records. This is often useful for implementing
> data retention policies and for freeing resources.

> Only workflow instances in the COMPLETED, FAILED, or TERMINATED state can be purged. If the workflow is in any other state, calling `purge` will return an error.

File: daprdocs/content/en/reference/api/workflow_api.md (hunk 17 above)
Comment (msfussell) [reply daprdocs/content/en/reference/api/workflow_api.md:78]: Either way I still do not understand how to use this API. See the conversation above.
Can we add
1) A description. "Raise an event from this workflow"
2) Describe when this is used. If this raises an event, how do you listen to this event and can a workflow also be trigger on the event. Please add a description here on the scenario both for workflow and how an app utilizes the event.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 28 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md:105]: Yes, this is much better, please update this.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 16 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Still confused here. between sending and registering for events 
 
So there is
1)A  language SDK specific way to register for events with `WaitForExternalEvent`you mention above ^^

2) And also there is this API
// Raise an event (an incoming purchase order) that your workflow will wait for. This returns the item waiting to be purchased.
await daprClient.RaiseWorkflowEventAsync(orderId, workflowComponent, workflowName, input, cts);

3) And then there is an HTTP way to send an event, but then says below that this will "Subscribe to external events" and "Await those tasks to block execution until the event is received.", so that description is wrong then?

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 29]
```diff
@@ -75,6 +75,113 @@ Code | Description
...
+`workflowComponentName` | Use `dapr` for Dapr Workflows
+`instanceId` | Unique value created for each run of a specific workflow
+
+### HTTP response codes
+
+Code | Description
+---- | -----------
+`202`  | Accepted
+`400`  | Request was malformed
+`500`  | Request formatted correctly, error in dapr code or underlying component
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/workflow_api.md]:
```suggestion
`500`  | Error in Dapr code or underlying component
```

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 30]
```diff
@@ -75,6 +75,113 @@ Code | Description
...
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/workflow_api.md]:
```suggestion
`500`  | Error in Dapr code or underlying component
```

**Review by msfussell — COMMENTED** [review]
More comments. Still do not understand the Raise an event usage.

**Review by msfussell — COMMENTED** [review]
More comments. Still do not understand the Raise an event usage.

**Review by msfussell — COMMENTED** [review]
More comments. Still do not understand the Raise an event usage.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 13 above)
Comment (cgillum) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: We could change the `GetWorkflow` API handler to return "PAUSED" instead of "SUSPENDED". I think we'd need to take this up with the Dapr maintainers as I expect there will be some angst about making changes this late in the milestone.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 16 above)
Comment (cgillum) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Oh, sorry, I wasn't paying close enough attention to the context of your question. Yes, this description is not accurate - there's nothing about this API that will do any kind of subscribing or awaiting. Only the workflow itself can do that.

Here's my proposal for how to describe this API:

*For workflow components that support subscribing to external events, such as the Dapr Workflow engine, you can use the following "raise event" API to deliver a named event to a specific workflow instance.*

```http
POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
```

File: daprdocs/content/en/reference/api/workflow_api.md (hunk 17 above)
Comment (cgillum) [reply daprdocs/content/en/reference/api/workflow_api.md:78]: I responded to the conversation above and recommended a change to how we describe his API. Hopefully that adds a bit more clarity.

One thing we need to mention is that the exact mechanism for subscribing to an event depends on the workflow component that you're using. Dapr Workflow has one way of subscribing to external events but other workflow components might have different ways.

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 31]
```diff
@@ -58,6 +70,43 @@ To terminate your workflow with an ID `12345678`, run:
 POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/terminate
 ```
 
+### Raise an event
+
+Run the following for your workflow to create "wait for external event" tasks that:
+
+1. Subscribe to external events.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]:
```suggestion
1. Workflow is registered for an external event  to happen in code
```

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md (hunk 16 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]: Yes, this is better.

**Comment by hhunter-ms** [comment]
@msfussell @cgillum @RyanLettieri thanks for the reviews! can you glance over it again to make sure I've addressed all your suggestions?

**Comment by RyanLettieri** [comment]
LGTM

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 32]
```diff
@@ -54,15 +65,49 @@ Note that workflow instance IDs can only contain alphanumeric characters, unders
...
+
+```http
+POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
+```
+
+> An `eventName` can be any function. 
+
+### Pause or resume a workflow
+
+To plan for down-time, wait for inputs, and more, you can pause and then resume a workflow. To stop a workflow with an ID `12345678` until triggered to resume, run:
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]:
```suggestion
To plan for down-time, wait for inputs, and more, you can pause and then resume a workflow. To pause a workflow with an ID `12345678` until triggered to resume, run:
```

File: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md [hunk 33]
```diff
@@ -54,15 +65,49 @@ Note that workflow instance IDs can only contain alphanumeric characters, unders
...
+
+```http
+POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/resume
+```
+
+### Purge a workflow 
+
+The purge API can be used to permanently delete workflow metadata from the underlying state store, including any stored inputs, outputs, and workflow history records. This is often useful for implementing data retention policies and for freeing resources.
+
+Only workflow instances in the COMPLETED, FAILED, or TERMINATED state can be purged. If the workflow is in any other state, calling purge will return an error.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md]:
```suggestion
Only workflow instances in the COMPLETED, FAILED, or TERMINATED state can be purged. If the workflow is in any other state, calling purge returns an error.
```

File: daprdocs/content/en/reference/api/workflow_api.md [hunk 34]
```diff
@@ -75,19 +75,133 @@ Code | Description
...
+---- | -----------
+`202`  | Accepted
+`400`  | Request was malformed
+`500`  | Error in Dapr code or underlying component
+
+### Response content
+
+None.
+
 ### Get workflow request
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/workflow_api.md]: Heading at the wrong level.
```suggestion
## Get workflow request
```

Can we also add the status into the table that can be returned. Notably `Running`, `Terminated` etc
```suggestion
### Get workflow request
```

**Review by msfussell — APPROVED** [review]
LGTM.
